core/src/main/java/jenkins/monitor/EndOfLifeAdminMonitor.java

```diff
@@ -0,0 +1,179 @@
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+import org.kohsuke.stapler.HttpRedirect;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.HttpResponses;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+/* Package protected so that it won't be visible outside this package */
+class EndOfLifeAdminMonitor extends AdministrativeMonitor {
+
+    /* URL of documentation that provides more details about operating system end of life */
+    static final String DOCUMENTATION_URL = "https://www.jenkins.io/redirect/operating-system-end-of-life";
+
+    /**
+     * Unique identifier of end of life admin monitor.
+     */
+    private final String identifier;
+
+    /**
+     * Display name of end of life admin monitor.
+     * Typically "End of life for (dependencyName)".
+     */
+    final String displayName;
+
+    /**
+     * Name of the dependency that will be end of life, like "Ubuntu 18.04"
+     */
+    final String dependencyName;
+
+    /**
+     * Date to begin displaying the end of life admin monitor, like 2023-01-31
+     */
+    final LocalDate beginDisplayDate;
+
+    /**
+     * Date that support ends, like 2023-05-31
+     */
+    final LocalDate endOfSupportDate;
+
+    /**
+     * URL with more information.
+     */
+    final String documentationURL = DOCUMENTATION_URL;
+
+    /**
+     * True if the dataPattern matched content of the file when constructor was
+     * called. Assumes the content of the file is not changing while Jenkins is
+     * running.
+     */
+    final boolean dataPatternMatched;
+
+    /**
+     * True if current date is after the end of support date
+     */
+    boolean unsupported;
+
+    /* Each end of life admin monitor needs to be separately enabled and disabled */
+    protected Boolean disabled;
+
+    EndOfLifeAdminMonitor(@NonNull String identifier,
+            @NonNull String dependencyName,
+            @NonNull LocalDate beginDisplayDate,
+            @NonNull LocalDate endOfSupportDate,
+            @NonNull File dataFile,
+            @NonNull Pattern dataPattern) {
+        super(identifier);
+        this.identifier = identifier;
+        this.disabled = SystemProperties.getBoolean(identifier + ".disabled", false);
+        this.dependencyName = dependencyName;
+        this.displayName = "End of life for " + dependencyName;
+        this.beginDisplayDate = beginDisplayDate;
+        this.endOfSupportDate = endOfSupportDate;
+        this.dataPatternMatched = patternMatched(dataFile, dataPattern);
+    }
+
+    private boolean patternMatched(File dataFile, Pattern dataPattern) {
+        boolean matched = false;
+        LOGGER.log(FINE, "Reading file {1} with monitor {0}", new Object[] {identifier, dataFile});
+        if (dataFile.isFile()) {
+            try (BufferedReader reader = new BufferedReader(Files.newBufferedReader(dataFile.toPath(), Charset.defaultCharset()))) {
```

Comment:
can we use utf-8 here instead of the default charset?

Reply:
Done in 89c39fac725aa05af5f46d6cb262d3d5e64cf1a3